Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm

```diff
@@ -750,4 +750,84 @@ function sendMessage(message) { \
+            objectStore.put(item, 'key'); \
+        }; \
+        request.onsuccess = function() { window.webkit.messageHandlers.testHandler.postMessage('success'); }; \
+        request.onerror = function() { window.webkit.messageHandlers.testHandler.postMessage('error'); }; \
+    </script>";
+    pid_t networkProcessIdentifier;
+    @autoreleasepool {
+        auto websiteDataStore = adoptNS([[WKWebsiteDataStore alloc] _initWithConfiguration:websiteDataStoreConfiguration.get()]);
+        auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+        [[configuration userContentController] addScriptMessageHandler:handler.get() name:@"testHandler"];
+        [configuration setWebsiteDataStore:websiteDataStore.get()];
+        auto webView = adoptNS([[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration.get()]);
+        [webView loadHTMLString:htmlString baseURL:[NSURL URLWithString:@"http://webkit.org"]];
+        TestWebKitAPI::Util::run(&receivedScriptMessage);
+        receivedScriptMessage = false;
+        EXPECT_WK_STREQ(@"success", [lastScriptMessage body]);
+        lastScriptMessage = nil;
+
+        [websiteDataStore fetchDataRecordsOfTypes:[NSSet setWithObject:WKWebsiteDataTypeIndexedDBDatabases] completionHandler:^(NSArray<WKWebsiteDataRecord *> *dataRecords) {
+            EXPECT_EQ((int)dataRecords.count, 1);
+            for (WKWebsiteDataRecord *record in dataRecords)
+                EXPECT_WK_STREQ([record displayName], @"webkit.org");
+            done = true;
+        }];
+        TestWebKitAPI::Util::run(&done);
+        done = false;
+
+        networkProcessIdentifier = [websiteDataStore.get() _networkProcessIdentifier];
+        EXPECT_NE(networkProcessIdentifier, 0);
+    }
+
+    // Wait until network process exits so we are sure new quota will be picked up.
+    while (!kill(networkProcessIdentifier, 0))
+        TestWebKitAPI::Util::spinRunLoop();
+
+    auto ratioNumber = [NSNumber numberWithFloat:0.000000001];
+    [websiteDataStoreConfiguration.get() setTotalQuotaRatio:ratioNumber];
+    [websiteDataStoreConfiguration.get() setOriginQuotaRatio:ratioNumber];
+    auto websiteDataStore = adoptNS([[WKWebsiteDataStore alloc] _initWithConfiguration:websiteDataStoreConfiguration.get()]);
+    auto configuration = adoptNS([[WKWebViewConfiguration alloc] init]);
+    [[configuration userContentController] addScriptMessageHandler:handler.get() name:@"testHandler"];
+    [configuration setWebsiteDataStore:websiteDataStore.get()];
+    auto webView = adoptNS([[WKWebView alloc] initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration.get()]);
+    [webView loadHTMLString:htmlString baseURL:[NSURL URLWithString:@"http://apple.com/"]];
+    TestWebKitAPI::Util::run(&receivedScriptMessage);
+    receivedScriptMessage = false;
+    EXPECT_WK_STREQ(@"error", [lastScriptMessage body]);
+
+    // Ensure webkit.org data is evicted.
```

Comment:
Can we test 3 origins, something like?
- load origin1
- load origin2
- reload origin1 to update LRU
- load origin 3 and verify origin2 is evicted.

And maybe the same test but with origin 2 page being opened so that origin 1 gets evicted?

Reply:
I can try: one thing a bit tricky is the quota is set based on disk space, so the test might require different values on different space. I guess I could introduce a mock disk space value.

Reply:
Updated test with 3 origins.